src/connection_string.ts

```diff
@@ -785,26 +775,792 @@ function parseURI(uri: string): { srv: boolean; url: URL; hosts: string[] } {
+  const valueString = String(value).toLowerCase();
+  if (TRUTHS.has(valueString)) return true;
+  if (FALSEHOODS.has(valueString)) return false;
+  throw new TypeError(`For ${name} Expected stringified boolean value, got: ${value}`);
+}
+
+function getInt(name: string, value: unknown): number {
+  if (typeof value === 'number') return Math.trunc(value);
+  const parsedValue = Number.parseInt(String(value), 10);
+  if (!Number.isNaN(parsedValue)) return parsedValue;
+  throw new TypeError(`Expected ${name} to be stringified int value, got: ${value}`);
+}
+
+function getUint(name: string, value: unknown): number {
+  const parsedValue = getInt(name, value);
+  if (parsedValue < 0) {
+    throw new TypeError(`${name} can only be a positive int value, got: ${value}`);
+  }
+  return parsedValue;
+}
+
+function toRecord(value: string): Record<string, any> {
+  const record = Object.create(null);
+  const keyValuePairs = value.split(',');
+  for (const keyValue of keyValuePairs) {
+    const [key, value] = keyValue.split(':');
+    record[key] = value;
+  }
+  return record;
+}
+
+const DEFAULT_PK_FACTORY = {
+  createPk(): ObjectId {
+    // We prefer not to rely on ObjectId having a createPk method
+    return new ObjectId();
+  }
+};
+
+const defaultOptions = new Map<string, unknown>(
+  ([
+    ['readPreference', ReadPreference.primary],
+    ['autoEncryption', {}],
+    ['compression', 'none'],
+    ['compressors', 'none'],
+    ['connectTimeoutMS', 30000],
+    ['dbName', 'test'],
+    ['directConnection', false],
+    ['driverInfo', {}],
+    ['forceServerObjectId', false],
+    ['heartbeatFrequencyMS', 10000],
+    ['keepAlive', true],
+    ['keepAliveInitialDelay', 120000],
+    ['localThresholdMS', 0],
+    ['logger', new Logger('MongoClient')],
+    ['maxIdleTimeMS', 0],
+    ['maxPoolSize', 100],
+    ['minPoolSize', 0],
+    ['monitorCommands', false],
+    ['noDelay', true],
+    ['numberOfRetries', 5],
+    ['pkFactory', DEFAULT_PK_FACTORY],
+    ['raw', false],
+    ['retryReads', true],
+    ['retryWrites', true],
+    ['serverSelectionTimeoutMS', 30000],
+    ['serverSelectionTryOnce', false],
+    ['socketTimeoutMS', 0],
+    ['waitQueueTimeoutMS', 0],
+    ['zlibCompressionLevel', 0]
+  ] as [string, any][]).map(([k, v]) => [k.toLowerCase(), v])
+);
+
 export function parseOptions(
   uri: string,
-  // eslint-disable-next-line @typescript-eslint/no-unused-vars
   options: MongoClientOptions = {}
 ): Readonly<MongoOptions> {
-  try {
-    // eslint-disable-next-line @typescript-eslint/no-unused-vars
-    const { srv, url, hosts } = parseURI(uri);
-    const mongoOptions: MongoOptions = ({ srv, hosts } as unknown) as MongoOptions;
-    // TODO(NODE-2699): option parse logic
-    return Object.freeze(mongoOptions);
-  } catch {
-    return Object.freeze({} as MongoOptions);
+  const { url, hosts, srv } = parseURI(uri);
+
+  // TODO(NODE-2704): Move back to test/tools/runner/config.js
+  options = { ...options };
```

Comment:
Is this line doing something useful?

Reply:
Yea it clones the options object so that on the next lines I can delete the host and port. Currently, the top level options make their way to socket construction so having a host/port option was overriding the host in the connection string. The ticket listed is the one where I'll go and make the options generated here actually used, so I'll be able to delete this since the host and port will be fetched from the right place.

I guess this leads to the question do we want to support host and port on the options object? and should it override the connection string even though currently uri options override object ones?